lib/hotel_block.rb

```diff
@@ -0,0 +1,96 @@
+require 'date'
+# # require_relative 'hotel.rb'
+
+module Booking
+  class Block
+
+    attr_reader :checkin_date, :checkout_date, :requested_rooms, :discounted_rate, :party_name, :room_block, :block_of_rooms, :all_rooms, :max_rooms, :room_numbers, :hotel, :all_reservations
+    
+    def initialize(checkin_date, checkout_date, requested_rooms, discounted_rate, hotel: nil)
+    @checkin_date = checkin_date
+    @checkout_date = checkout_date 
+    @block_of_rooms ||= []
+    @hotel ||= hotel
+    @available_rooms = @hotel.check_availability(checkin_date, checkout_date)
+    @all_reservations = @hotel.all_reservations
+    @discounted_rate = discounted_rate
+
+    requested_rooms.each do |room|
+      if @block_of_rooms.include?(room)
+        raise ArgumentError, "That room is already in a block."
+      elsif !@available_rooms.include?(room)
```

Comment:
I may argue that the logic checking if the requested rooms are valid and are included in the available rooms does not belong in the `initialize` method of a new `Block`